tools/palette.py

```diff
@@ -0,0 +1,645 @@
+def parse_gpl_pal(input: UnparsedPaletteData) -> ParsedPaletteData:
+    output = ParsedPaletteData()
+    output.palette_type = input.palette_type
+
+    with open(input.args.palette, "rt") as f:
+        lines = f.readlines()
+
+        for index, line in enumerate(lines, start=0):
+            line = line.strip()
+            
+            # Make sure this is a valid GPL or PAL format file
+            if index == 0:
+                if line.lower().find("gimp palette") == 0:
+                    output.palette_type = PaletteType.GPL
+                    continue
+                elif line.lower().find("jasc-pal") == 0:
+                    output.palette_type = PaletteType.PAL
+                    continue
+                elif input.args.force == False:
+                    raise Exception(f"ERROR: Invalid or malformed palette file passed to palette.py! {input.args.palette}")
+
+            # Parse GPL
+            if output.palette_type == PaletteType.GPL:
+                # Name may not be included
+                if line.lower().find("name:") != -1:
+                    nameSplit = line.split(":")
+                    output.palette_name = nameSplit[-1]
+                    continue
+                # Color count is common, but not guaranteed as part of spec
+                if line.lower().find("colors:") != -1:
+                    color_countSplit = line.split(":")
+                    output.color_count = int(color_countSplit[-1])
+                    continue
+
+                # Ignore other commented lines
+                if line.find("#") != -1:
+                    continue
+            
+            # Parse PAL
+            else:
+                if index == 1:
+                    continue
+                # Color count is part of spec
+                if index == 2:
+                    output.color_count = int(line)
+                    continue
+            
+            columns = line.split()
+            if len(columns) >= 3:
+                # GPL and PAL files do not support transparency, so define alpha 255 or 0
+                # based on the '--alpha' argument
+                a = 255
+                if len(output.palette_colors) == input.args.alpha:
+                    a = 0
+                output.palette_colors.append([int(columns[0]), int(columns[1]), int(columns[2]), a])
+
+    if output.color_count != -1 and len(output.palette_colors) != output.color_count and input.args.force == False:
+        raise Exception(f"ERROR: Parsed color count does not match file declaration! {len(output.palette_colors)} != {output.color_count}")
+    elif output.color_count == -1:
+        output.color_count = len(output.palette_colors)
+
+    # Validate color count is either 16 or 256
+    if output.color_count != 16 and output.color_count != 256:
+        raise Exception(f"ERROR: Parsed color count does not align with supported NAOMI palette sizes! {output.color_count} != 16 or 256")
```

Comment:
I wonder if it makes sense here to just fill the rest with an obnoxious color like (0x255, 0, 0x255). Presumably, somebody converting a palette that doesn't have all 16/256 colors is working with textures that simply do not have that many colors in them.